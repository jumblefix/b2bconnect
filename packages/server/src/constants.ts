export const REDIS_SESSION_PREFIX = 'sess:';
export const USER_SESSION_PREFIX = 'userSids:';
export const FORGOT_PASSWORD_PREFIX = 'forgotPassword:';
export const ITEMS_PER_PAGE = 10;

export enum Env {
  development = 'development',
  test = 'test',
  production = 'production',
}

export enum TokenTypes {
  confirm = 'confirm',
  reset = 'reset',
}
