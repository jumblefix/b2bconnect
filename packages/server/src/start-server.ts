// tslint:disable-next-line:no-var-requires
require('dotenv-safe').config();
import * as connectRedis from 'connect-redis';
import * as ExpressRateLimit from 'express-rate-limit';
import * as ExpressSession from 'express-session';
import { GraphQLServer } from 'graphql-yoga';
import * as helmet from 'helmet';
import * as IORedis from 'ioredis';
import * as RateLimitRedis from 'rate-limit-redis';
// tslint:disable-next-line:no-import-side-effect
import 'reflect-metadata';
import { Env, REDIS_SESSION_PREFIX, TokenTypes } from './constants';
import { seeder } from './seeder';
import { connectDb } from './utils/connect-db';
import { createDb } from './utils/create-db';
import { genSchema } from './utils/schema-utils';
import { confirmEmail, resetPassword } from './utils/user-utils';
import { formatError } from './utils/utils';

const RedisStore = connectRedis(ExpressSession as any);

export const redis = new IORedis();

export async function startServer() {
  if (process.env.NODE_ENV === Env.test) {
    await redis.flushall();
  }

  await createDb();

  const connection = await connectDb();

  if (process.env.NODE_ENV === Env.production) {
    await connection.runMigrations();
  }

  if (process.env.NODE_ENV !== Env.test) {
    await seeder(connection);
  }

  const server = new GraphQLServer({
    schema: genSchema(),
    context: ({ request }) => ({
      redis,
      url: `${request.protocol}://${request.get('host')}`,
      session: request.session,
      db: connection,
      req: request,
    }),
  });

  const { express } = server;

  express.use(helmet());

  express.use(
    new ExpressRateLimit({
      store: new RateLimitRedis({
        client: redis,
      }),
      windowMs: 15 * 60 * 1000, // 15 minutes
      max: 100, // limit each IP to 100 requests per windowMs
      delayMs: 0, // disable delaying - full speed until the max limit is reached
    }),
  );

  express.use(
    ExpressSession({
      store: new RedisStore({
        client: redis as any,
        prefix: REDIS_SESSION_PREFIX,
      }),
      name: 'qid',
      secret: process.env.SESSION_SECRET as string,
      resave: false,
      saveUninitialized: false,
      cookie: {
        httpOnly: true,
        // secure: process.env.NODE_ENV === Env.production,
        secure: false,
        maxAge: 1000 * 60 * 60 * 24 * 7, // 7 days
      },
    }),
  );

  express.get('/', (_, res) => res.json({ message: 'pong' }));
  express.get(`/${TokenTypes.confirm}/:id`, confirmEmail);
  express.get(`/${TokenTypes.reset}/:id`, resetPassword);

  const { NODE_ENV } = process.env;

  return server.start(
    {
      port: NODE_ENV === Env.test ? 4001 : 4000,
      formatError: (err: any) => formatError(err),
      playground: NODE_ENV === Env.production ? false : '/playground',
      endpoint: '/api',
    },
    ({ port }) => console.log('localhost:' + port),
  );
}
