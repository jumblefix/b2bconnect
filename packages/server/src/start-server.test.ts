import axios from 'axios';
import { Server } from 'net';
import { TokenTypes } from './constants';
import { startServer } from './start-server';

let server: Server;
beforeAll(async () => {
  server = await startServer();
});

afterAll(async () => {
  if (server) {
    await server.close();
  }
});

const { TEST_HOST } = process.env;

describe('start-server', () => {
  it('should start', async () => {
    const { data } = await axios.get(`${TEST_HOST}/`);
    expect(data).toMatchSnapshot();
  });

  it('confirm', async () => {
    const { data } = await axios.get(`${TEST_HOST}/${TokenTypes.confirm}/abc`);
    expect(data).toMatchSnapshot();
  });

  it('reset', async () => {
    const { data } = await axios.get(`${TEST_HOST}/${TokenTypes.reset}/abc`);
    expect(data).toMatchSnapshot();
  });
});
