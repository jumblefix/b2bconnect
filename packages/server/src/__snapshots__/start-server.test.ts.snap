// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`start-server confirm 1`] = `
Object {
  "message": "invalid or expired register token",
}
`;

exports[`start-server reset 1`] = `
Object {
  "message": "invalid or expired reset token",
}
`;

exports[`start-server should start 1`] = `
Object {
  "message": "pong",
}
`;
